--- a/backend/ipp.c
+++ b/backend/ipp.c
@@ -1736,7 +1736,10 @@
 	ippAddBoolean(request, IPP_TAG_OPERATION, "last-document",
         	      (i + 1) >= num_files);
 
-	if (document_format)
+	if (num_files > 1)
+	  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_MIMETYPE,
+		       "document-format", NULL, "application/octet-stream");
+	else if (document_format)
 	  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_MIMETYPE,
 		       "document-format", NULL, document_format);
 
