--- a/PPD/Generic-OAKT_Printer.ppd
+++ b/PPD/Generic-OAKT_Printer.ppd
@@ -33,7 +33,8 @@
 *cupsVersion:	1.0
 *cupsManualCopies: True
 *cupsModelNumber:  2
-*cupsFilter:	"application/vnd.cups-postscript 0 foomatic-rip"
+*cupsFilter: "application/vnd.cups-postscript 100 foomatic-rip"
+*cupsFilter: "application/vnd.cups-pdf 0 foomatic-rip"
 *%pprRIP:        foomatic-rip other
 *ModelName:     "Generic OAKT Printer"
 *ShortNickName: "Generic OAKT Printer foo2oak"
--- a/PPD/Generic-ZjStream_Printer.ppd
+++ b/PPD/Generic-ZjStream_Printer.ppd
@@ -33,7 +33,8 @@
 *cupsVersion:	1.0
 *cupsManualCopies: True
 *cupsModelNumber:  2
-*cupsFilter:	"application/vnd.cups-postscript 0 foomatic-rip"
+*cupsFilter: "application/vnd.cups-postscript 100 foomatic-rip"
+*cupsFilter: "application/vnd.cups-pdf 0 foomatic-rip"
 *%pprRIP:        foomatic-rip other
 *ModelName:     "Generic ZjStream Printer"
 *ShortNickName: "Generic ZjStre. Printer foo2zjs"
